sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSchemaInferenceSuite.scala

```diff
@@ -0,0 +1,192 @@
+      lowercaseSchema,
+      Seq("partcol1", "partcol2"))
+    client.createTable(catalogTable, true)
+
+    val partitions = (0 until NUM_RECORDS).map(hiveCatalogPartition(dir.getAbsolutePath, _)).toSeq
+    client.createPartitions(DATABASE, tableName, partitions, true)
+  }
+
+  // Create a test table used for a single unit test, with data stored in the specified directory.
+  private def withTestTable(dir: File)(f: File => Unit): Unit = {
+    setupCaseSensitiveTable(TEST_TABLE_NAME, dir)
+    try f(dir) finally spark.sql(s"DROP TABLE IF EXISTS $TEST_TABLE_NAME")
+  }
+
+  private val inferenceKey = SQLConf.HIVE_CASE_SENSITIVE_INFERENCE.key
+
+  test("Schema should be inferred and written to table properties when INFER_AND_SAVE is " +
+    "specified") {
+    withSQLConf(inferenceKey -> HiveCaseSensitiveInferenceMode.INFER_AND_SAVE.toString) {
+      withTempDir { dir =>
+        withTestTable(dir) { dir =>
+          assert(spark.sql(FIELD_QUERY).count == NUM_RECORDS)
+          assert(spark.sql(PARTITION_COLUMN_QUERY).count == NUM_RECORDS)
+          // Test that the case-sensitive schema was storied as a table property after inference
+          assert(spark.sql(SELECT_ALL_QUERY).schema == caseSensitiveSchema)
+
+          // Verify the catalog table now contains the udpated schema and properties
+          val catalogTable = externalCatalog.getTable(DATABASE, TEST_TABLE_NAME)
+          assert(catalogTable.schemaPreservesCase == true)
+          assert(catalogTable.schema == caseSensitiveSchema)
+          assert(catalogTable.partitionColumnNames == Seq("partcol1", "partcol2"))
+          val rawTable = externalCatalog.client.getTable(DATABASE, TEST_TABLE_NAME)
+          assert(rawTable.properties.contains(HiveExternalCatalog.DATASOURCE_SCHEMA_NUMPARTS))
+        }
+      }
+    }
+  }
+
+  test("Schema should be inferred but not stored when INFER_ONLY is specified") {
+    withSQLConf(inferenceKey -> HiveCaseSensitiveInferenceMode.INFER_ONLY.toString) {
+      withTempDir { dir =>
+        withTestTable(dir) { dir =>
+          val originalTable = externalCatalog.getTable(DATABASE, TEST_TABLE_NAME)
+          val existingSchema = spark.sql(SELECT_ALL_QUERY).schema
+          assert(spark.sql(FIELD_QUERY).count == NUM_RECORDS)
+          assert(spark.sql(PARTITION_COLUMN_QUERY).count == NUM_RECORDS)
+          assert(spark.sql(SELECT_ALL_QUERY).schema == existingSchema)
+          // Catalog table shouldn't be altered
+          assert(externalCatalog.getTable(DATABASE, TEST_TABLE_NAME) == originalTable)
+        }
+      }
+    }
+  }
+
+  test("Schema should not be inferred when NEVER_INFER is specified") {
+    withSQLConf(inferenceKey -> HiveCaseSensitiveInferenceMode.NEVER_INFER.toString) {
+      withTempDir { dir =>
+        withTestTable(dir) { dir =>
+          val originalTable = externalCatalog.getTable(DATABASE, TEST_TABLE_NAME)
+          // Only check the schema returned by SELECT * as other queries will break
```

Comment:
As mentioned in #16797 this issue actually won't cause exceptions, at least for Parquet data. The queries will simply return 0 results due to ParquetReadSupport using case-sensitive field resolution. If enabled, any pushed-down filter containing a case-sensitive field will also return 0 results since the lowercase filter field name won't match the case-sensitive Parquet column name.

I'll put some thought towards whether this test can be made more robust in other ways.